### DISCUSSION TITLE
Tcp port forwarding with ssl

### DISCUSSION THREAD

**Discussion by MinitorMHS** [NginxProxyManager/nginx-proxy-manager#4207] [Q&A] [answered]
I'm trying to setup ngnix to use my subdomain and port 443 to forward requests to port 3000 which is a tcp port , for http ports it's easy to do this but did anyone do this before or can help me achieve this?

**Comment by shalafi99** [comment]
Hi, the protocol involved (https/http or another) would be more relevant than the actual TCP port involved while deciding on what to set up inside NPM.

You say you want NPM to listen for https traffic (TCP 443) using that subdomain of yours and for that, you will add a TLS certificate into NPM (either as an external certificate or using NPM's feature to request a new certificate).
And the traffic it receives will be sent to a destination listening on port TCP 3000.
All above is correct understanding?

That destination host/app/service which is listening on TCP 3000, is it using http or https?

**Reply by MinitorMHS** [reply]
Yeah you understood correctly, I want npm to listen to my subdomain on port 443 and forward the requests to tcp port 3000 which is tls , but about ssl certificates I can configure my application to use certs or neither npm so it's not really important cause I can do it both sides

**Reply by shalafi99** [reply]
So, for that it would be a proxy host.
You mentioned tls being used at TCP 3000 so https for the 'scheme' drop-down.
And for the NPM listener, what you select in the "SSL" tab determines http or https.

![image](https://github.com/user-attachments/assets/42c919c1-49a1-4445-a808-b9e990fb13b7)

**Reply by MinitorMHS** [reply]
I selected https but that's where my problem is cause it won't work, as I searched before opening this issue nginx can only handle tcp ports in streams section but I don't really know if I'm right or not
And yes as you mentioned I'm running a proxy on port 3000

**Reply by shalafi99** [reply, marked as answer]
You would use a stream host only for non http/https traffic, as far as I know.
_(That's why I began mentioning the protocol in use is more relevant)_

A couple of examples I have in my environment:
![image](https://github.com/user-attachments/assets/140c50ef-442f-49ae-b53e-e7cad889273f)
--> traffic forwarded to a host running https on TCP 8443

![image](https://github.com/user-attachments/assets/9ddc21bb-23d0-4bc5-83fe-2b073621f9fb)
--> traffic forwarded to the NPM container's own web gui port, http on TCP 81
_(upon posting, realized websockets should not be turned on here, gone now)_


For both, it might as well have been TCP 3000 instead of what is now (in case I changed the service's port to that)

When we think just TCP it's the Transport Layer (OSI Layer4), but when we add to that the knowledge it's http/https we have the Application Layer (OSI Layer7) thrown into the mix.

NPM Stream host = Layer4
NPM Proxy host = Layer7
And that's so no matter the actual TCP port in use.

**Reply by MinitorMHS** [reply]
Oh thanks for your helpful explanation, and one more question, what if my application is not running in docker? What should I put in "Forward hostname/ip"

**Reply by shalafi99** [reply]
You might have a couple options there, mostly depending on how the service is setup and where it is located inside the network in relation to docker/NPM.

Which one below would you use to reach the proxy at TCP 3000 directly? (=no NPM involved)

IP (maybe because it's a static one)
a fully qualified domain name (maybe because it has a dynamic IP, or because TLS being involved requires that)
localhost/127.0.0.1 (because it's running directly where you access it from)

I suppose it only gets tricky when it is running inside the same host where docker is installed and listening only on the loopback address and no other interfaces. Might that be your case?

**Reply by MinitorMHS** [reply]
Could you please dm me on telegram? [https://t.me/minitor]()

**Reply by MinitorMHS** [reply]
So could you please give me a TON wallet address, I need to pay you something, it's not much but just for your time

**Reply by shalafi99** [reply]
There is no need for that, but thanks.
I'm just contributing some of my time to provide help if/where I can, of my free will and without expecting a material return for it.

_(As for the negative answer before, taking the discussion outside of this platform would prevent others from benefiting from what is shared, deviating from how I aim to contribute)_

My last question did not imply you would have to share sensitive / actual information from your own environment, just to get a better idea of how it is set up and thus not have to provide an answer either too broad/generic or more extensive than actually needed.

**Reply by MinitorMHS** [reply]
No I didn't mean that , I solved my problem, already I found that for using local host address in npm I should add ```network_mode: host``` to docker-compose file so now npm will listen to 443 and forward requests to 3000 thanks to your full explanation of streams and hosts , by the way since I'm running a proxy service (v2ray based on [xtls.github.io]()) and the protocol is websocket I also enabled websocket in npm I just wanted you to know this cause you said I should disable websocket in npm 
Again I want to say thank you for your help and if available give me an email or wallet or something so I can pay you something for your troubles 
Also I'm using [github.com/alireza0/s-ui]() for proxy panel , try it you won't regret it cause I was you've also done some contribution on wg-easy

**Reply by shalafi99** [reply]
Great to hear you got it working.
And again, thanks for the offer but no payment required.

Oh, my comment regarding websocket support was specific for NPM's web admin interface, its usage is case by case depending on what the service behind NPM is used for. NPM's own admin panel does not need it (the example I provided)

Since you are all set, for a wrap-up you could mark this thread as closed.
(But not sure if/where to pick any particular reply as the answer would be applicable, though)

**Reply by MinitorMHS** [reply]
Of course 
Thank you Shalafi99❤️
